spec/models/calculated_handover_date_spec.rb

```diff
@@ -0,0 +1,161 @@
+    )
+
+    record.reload
+    expect(record.start_date).to be_nil
+    expect(record.handover_date).to be_nil
+    expect(record.reason).to eq('COM Responsibility')
+  end
+
+  describe "when nomis_offender_id is set but an associated case information record doesn't exist" do
+    subject {
+      build(:calculated_handover_date,
+            case_information: nil,
+            nomis_offender_id: "A1234BC"
+      )
+    }
+
+    it 'is not valid' do
+      expect(subject.valid?).to be(false)
+      expect(subject.save).to be(false)
+    end
+  end
+
+  describe '#recalculate_for(offender)' do
+    let(:offender) { build(:offender) }
+    let!(:case_info) { create(:case_information, :nps, nomis_offender_id: offender.offender_no) }
+
+    before { offender.load_case_information(case_info) }
+
+    describe "when calculated handover dates don't exist yet for the offender" do
+      it 'creates a new record' do
+        expect {
+          described_class.recalculate_for(offender)
+        }.to change(described_class, :count).from(0).to(1)
+
+        record = described_class.find_by(nomis_offender_id: offender.offender_no)
+        expect(record.start_date).to eq(offender.handover_start_date)
+        expect(record.handover_date).to eq(offender.responsibility_handover_date)
+        expect(record.reason).to eq(offender.handover_reason)
+      end
+    end
+
+    describe 'when calculated handover dates already exist for the offender' do
+      let!(:existing_record) {
+        create(:calculated_handover_date,
+               case_information: case_info,
+               start_date: existing_start_date,
+               handover_date: existing_handover_date,
+               reason: existing_reason
+        )
+      }
+
+      before do
+        expect(existing_record.start_date).to eq(existing_start_date)
+        expect(existing_record.handover_date).to eq(existing_handover_date)
+        expect(existing_record.reason).to eq(existing_reason)
+      end
+
+      describe 'when the dates have changed' do
+        let(:existing_start_date) { today + 1.week }
+        let(:existing_handover_date) { existing_start_date + 7.months }
+        let(:existing_reason) { 'CRC Case' }
+
+        it 'updates the existing record' do
+          described_class.recalculate_for(offender)
+
+          existing_record.reload
+          expect(existing_record.start_date).to eq(offender.handover_start_date)
+          expect(existing_record.handover_date).to eq(offender.responsibility_handover_date)
+          expect(existing_record.reason).to eq(offender.handover_reason)
+        end
+      end
+
+      describe "when the dates haven't changed" do
+        let(:existing_start_date) { offender.handover_start_date }
+        let(:existing_handover_date) { offender.responsibility_handover_date }
+        let(:existing_reason) { offender.handover_reason }
+
+        it "does nothing" do
+          old_updated_at = existing_record.updated_at
+
```

Comment:
tests can set updated_at, so there's no need to sleep - simply check that it hasn't changed

Reply:
I was following the approach used in the CaseInformation spec:
https://github.com/ministryofjustice/offender-management-allocation-manager/blob/72f6199ce652d8866bdd0daa06e2d272cfad9845/spec/models/case_information_spec.rb#L6-L11

But it did feel a little weird to sleep in the tests.

I've just changed it to use time travel instead. What do you think of this approach @starswan?